### ISSUE TITLE
question: what is the "right way" to use netmap from another linux module?

### ISSUE THREAD

**Issue by linoferrentino** [luigirizzo/netmap#609]
Hi all,

I am trying to write another module which depends on netmap.

That module should intercept some system call related to sockets and redirect them
to a netmap enabled interface. I have already done a library that does that in
user space and it works (using LD_PRELOAD) but not all the system calls
of a generic executable are using the glibc wrapper.

So I would like to move the code to a kernel module, but what is the
"right way" to do? For example how could I open a netmap pipe from a
kernel module?

Suppose that the interface "eth0" has already put in netmap mode from
a user application, we have only to open one end of the pipe from the
module.

Thanks.

**Comment by vmaffione** [comment]
From you kernel module you can call the functions exported by the kernel image or by other kernel modules such as netmap (look for EXPORT_SYMBOL and EXPORT_SYMBOL_GPL).
However, I don't think there are enough exported function for what you want to do.
You may change netmap to export everything you need, but there is probably a better approach.

You  could write a kernel module that receives netmap file descriptors from user-space, and access the netmap rings from user-space. For synchronization, you can use something like `ksys_ioctl()`, `do_vfs_ioctl()`, `vfs_poll`, all accessible from kernel-space. You would still need an user-space daemon that opens the netmap ports to get the file descriptors.
ksys_ioctl() or

**Comment by linoferrentino** [comment]
Thanks, I will try.

**Comment by linoferrentino** [comment]
Hi,
just in case someone else might have to do this...

I have managed to call netmap from another kernel module. The steps  that I have done are:

Exported some symbols

```
+EXPORT_SYMBOL(netmap_priv_new);
+EXPORT_SYMBOL(netmap_global_lock);
+EXPORT_SYMBOL(netmap_dtor);
+EXPORT_SYMBOL(netmap_ioctl);
```

My module then creates a device and a netmap device internal to it, something like


```
        /* XXX I don't like to use the netmap lock here: fix! */
        NMG_LOCK();
        filp->nm_priv = netmap_priv_new();
        if (filp->nm_priv == NULL) {
                kfree(filp);
                printk(KERN_ERR "out of memory allocating netmap module");
                res = -ENOMEM;
                goto end;
        }

 end:
        NMG_UNLOCK();

```

Then I create the actual device (a pipe) calling directly the ioctl,

This is a test:

```
        /* Now we request the netmap interface. */
        hdr->nr_version = NETMAP_API;   /* new API */
        strcpy(hdr->nr_name, "eth0}0");

        hdr->nr_body = (uintptr_t) req;
        hdr->nr_reqtype = NETMAP_REQ_REGISTER;

        req->nr_ringid = 0;
        req->nr_mode = NR_REG_ALL_NIC;

        res = netmap_ioctl(filp->nm_priv, NIOCCTRL, (caddr_t) hdr, NULL,
                           /*nr_body_is_user= */ 0);

```

Then to send a packet I use the same code as in userspace, but I use
the macro NMB to get the address, because now we are on the other side

```
                        slot = &ring->slot[ring->head];
                        printk("The slot is %p, its idx is %d",
                               slot, slot->buf_idx);

                        pdst = NMB(na, slot);
                        printk("The pointer is %p", pdst);
                        memcpy(pdst, packet, sizeof(packet));


```

Then the send is done as usual


```
                        res =
                            netmap_ioctl(priv->filp->nm_priv, NIOCTXSYNC, NULL,
                                         NULL,
                                         /*nr_body_is_user= */ 0);

```

the receiving part is done in this way.

the module will wait on the rxring queue (the 'si' member).

```
                        htail = rkring->pipe_tail;
                        res = wait_event_interruptible_timeout
                            (rkring->si, htail != rkring->pipe_tail, 5000);
```

then the netmap module will notify myself when the pipe_tail changes
and I will process the new packets.

For now it is working, even if a bit ugly because I use
some internal structures and fields...

but it's a starting point.

Thanks

**Comment by vmaffione** [comment]
Ok, thanks. I still think the right solution is the one where you use user-space `nm_open` or similar.
